File: backend/src/routes/ea-review-logs.ts
import { Router, Request, Response } from 'express';
import { PrismaClient } from '@prisma/client';
import { getPaginationParams, buildPaginatedResponse } from '../middleware/pagination';

const prisma = new PrismaClient();
const router = Router();

function mapLog(l: any) {
  return {
    id:        l.id,
    comment:   l.comment,
    createdAt: l.create_at,
    operator:  l.operator,
    operation: l.operation,
    projectId: l.project_id,
  };
}

router.get('/', async (req: Request, res: Response) => {
  try {
    const { page, pageSize, sortField, sortOrder, skip } = getPaginationParams(req);
    const { projectId, operator } = req.query;

    const where: any = {};
    if (projectId) where.project_id = projectId as string;
    if (operator)  where.operator   = { contains: operator as string, mode: 'insensitive' };

    const fieldMap: Record<string, string> = { createdAt: 'create_at' };
    const dbSortField = sortField ? (fieldMap[sortField as string] || (sortField as string)) : 'create_at';
    const orderBy: any = { [dbSortField]: sortOrder || 'desc' };

    const [data, total] = await Promise.all([
      prisma.eam_review_log.findMany({ where, orderBy, skip, take: pageSize }),
      prisma.eam_review_log.count({ where }),
    ]);

    res.json(buildPaginatedResponse(data.map(mapLog), total, page, pageSize));
  } catch (error) {
    console.error(error);
    res.status(500).json({ error: 'Failed to fetch EA review logs' });
  }
});

export { router as eaReviewLogRoutes };

File: backend/app/routers/ea_review_logs.py
"""EA review log endpoints."""
from fastapi import APIRouter, Depends, Query
from fastapi.responses import JSONResponse
from sqlalchemy.ext.asyncio import AsyncSession
from sqlalchemy import text

from app.database import get_db
from app.utils.pagination import PaginationParams, paginated_response

from app.auth import require_permission, require_role, Role

router = APIRouter()

FIELD_MAP = {
    "createdAt": "create_at",
}

ALLOWED_SORT_COLUMNS = {
    "id", "comment", "create_at", "operator", "operation", "project_id",
}


def _map_log(row: dict) -> dict:
    return {
        "id": row["id"],
        "comment": row["comment"],
        "createdAt": row["create_at"].isoformat() if row["create_at"] else None,
        "operator": row["operator"],
        "operation": row["operation"],
        "projectId": row["project_id"],
    }


@router.get("", dependencies=[Depends(require_permission("ea_review_log", "read"))])
async def get_ea_review_logs(
    pag: PaginationParams = Depends(),
    projectId: str | None = Query(None),
    operator: str | None = Query(None),
    db: AsyncSession = Depends(get_db),
):
    try:
        conditions: list[str] = []
        params: dict = {}

        if projectId:
            conditions.append("project_id = :project_id")
            params["project_id"] = projectId
        if operator:
            conditions.append("operator ILIKE :operator")
            params["operator"] = f"%{operator}%"

        where_clause = " AND ".join(conditions) if conditions else "1=1"

        # Resolve sort column
        db_sort_field = FIELD_MAP.get(pag.sort_field, pag.sort_field) if pag.sort_field else "create_at"
        if db_sort_field not in ALLOWED_SORT_COLUMNS:
            db_sort_field = "create_at"
        sort_order = pag.sort_order if pag.sort_order in ("asc", "desc") else "desc"

        # Count query
        count_result = await db.execute(
            text(f"SELECT COUNT(*) AS cnt FROM eam.eam_review_log WHERE {where_clause}"),
            params,
        )
        total = count_result.scalar_one()

        # Data query
        data_result = await db.execute(
            text(
                f"SELECT * FROM eam.eam_review_log WHERE {where_clause} "
                f"ORDER BY {db_sort_field} {sort_order} "
                f"LIMIT :limit OFFSET :offset"
            ),
            {**params, "limit": pag.page_size, "offset": pag.offset},
        )
        rows = data_result.mappings().all()

        return paginated_response(
            [_map_log(dict(r)) for r in rows],
            total,
            pag.page,
            pag.page_size,
        )
    except Exception as e:
        print(f"Error: {e}")
        return JSONResponse(status_code=500, content={"error": "Failed to fetch EA review logs"})
